internal/controller/stages/stages.go

```diff
@@ -647,198 +652,196 @@ func (r *reconciler) syncNormalStage(
 	status.ObservedGeneration = stage.Generation
 	status.Health = nil
 
-	if status.CurrentFreight == nil {
+	if currentFreight := status.FreightHistory.Current(); currentFreight == nil || len(currentFreight.Items) == 0 {
 		status.Phase = kargoapi.StagePhaseNotApplicable
 		logger.Debug(
 			"Stage has no current Freight; no health checks or verification to perform",
 		)
 	} else {
-		freightLogger := logger.WithValues("freight", status.CurrentFreight.Name)
-		shouldRecordFreightVerificationEvent := false
-
-		// Push the latest state of the current Freight to the history at the
-		// end of each reconciliation loop.
-		defer func() {
-			status.History.UpdateOrPush(*status.CurrentFreight)
-		}()
-
-		// Always check the health of the Argo CD Applications associated with the
-		// Stage. This is regardless of the phase of the Stage, as the health of the
-		// Argo CD Applications is always relevant.
-		if status.Health = r.appHealth.EvaluateHealth(
-			ctx,
-			*status.CurrentFreight,
-			stage.Spec.PromotionMechanisms.ArgoCDAppUpdates,
-		); status.Health != nil {
-			freightLogger.WithValues("health", status.Health.Status).Debug("Stage health assessed")
-		} else {
-			freightLogger.Debug("Stage health deemed not applicable")
-		}
-
-		if stage.Spec.Verification != nil {
-			// Update the verification history with the current verification info.
-			// NOTE: We do this regardless of the phase of the verification process
-			// and before potentially creating a new AnalysisRun to ensure we add
-			// any previous verification attempt which may have been recorded
-			// before we started tracking history.
-			if status.CurrentFreight.VerificationInfo != nil {
-				status.CurrentFreight.VerificationHistory.UpdateOrPush(*status.CurrentFreight.VerificationInfo)
+		for _, freight := range currentFreight.Items {
+			freightLogger := logger.WithValues("freight", status.FreightHistory.Current())
```

Comment:
Are these k/v pairs correct? `status.FreightHistory.Current()` is a struct and is probably going to dump a lot of noise into the logs.

Reply:
Good catch, this does indeed require changes.